### ISSUE TITLE
Proposal: Anonymous function literals with function signature inference

### ISSUE THREAD

**Issue by Rocknest** [ziglang/zig#4170]
## Rationale ##
In the discussion of '#1717 function expressions' there are a lot of comments that touch a topic of function signature inference:
https://github.com/ziglang/zig/issues/1717#issuecomment-444200663 https://github.com/ziglang/zig/issues/1717#issuecomment-514536941 https://github.com/ziglang/zig/issues/1717#issuecomment-565284947 https://github.com/ziglang/zig/issues/1717#issuecomment-573472295. Which is a natural use case for anonymous function expressions (classis example is an argument to 'sort'), however as stated by @hryx https://github.com/ziglang/zig/issues/1717#issuecomment-573535638 this is not a goal of that proposal, so this use case will have to be solved anyway, so thats why i'm creating this proposal.

This proposal is compatible with #1717 (which is 'accepted' at the time of writing) but in the substance makes it a bit redundant leaving the controversial [[1]](https://github.com/ziglang/zig/issues/1717#issuecomment-573429047) [[2]](https://github.com/ziglang/zig/issues/1717#issuecomment-573472295) 'syntactic consistency among all statements which bind something to an identifier' as the main difference.

Closures are non-goal.

## The proposal ##
Add an ability to declare a function inside of an expression. The types of the arguments of the function should be infered from the context, the same applies to the return type.

Possible syntax:
```zig
.|a, b| {
    // function body
}
```
An opening dot is a convention established by anon enum literals, structs, etc.
Parameters enclosed by `| |` instead of parentheses is also present in the language (eg. loops).

Such expressions should be coercable to the function type that is expected in the declaring scope.
```zig
const f: fn (i32) bool = .|a| {
    return (a < 4);
};

var f2: fn (i32) bool = if (condition) .|x| {
    return (x < 4);
} else .|x| {
    return (x == 54);
};
```
&nbsp;
Ambiguous expressions should probably be compile errors:
```zig
const a = .|| { return error.Unlucky; };
// @TypeOf(a) == fn () !void

const lessThan = .|a, b| {
    return a < b;
};
// error: ambiguous
// OR
// @TypeOf(lessThan) == fn (var, var) bool

fn foo() void {
    const incr = .|x| {
        return x + 1;
    };

    warn("woah {}\n", .{ incr(4) }); // Ok?
};
```
&nbsp;
Some examples:
```zig
pub fn sort(comptime T: type, arr: []T, f: fn (T, T) bool) {
    // ...
};

pub fn main() void {
    var letters = []u8 {'g', 'e', 'r', 'm', 'a', 'n', 'i', 'u', 'm'};

    sort(u8, letters, .|a, b| {
        return a < b;
    });
};
```
```zig
(.|| {
    std.debug.warn("almost js", .{});
    // what will @This() return?
})();
```

## Expression expressions ##
Naming these gets complicated:
```zig
sort(u8, letters, .|a, b| => (a < b));
```
Basically a shortcut for one line function that return some expression. These are not part of this proposal.

**Comment by mikdusan** [comment]
What if functions vs. function-type is explicit:

```zig
const f: Fn(i32) bool = fn(a) {
    return (a < 4);
};

var f2: Fn(i32) bool = if (condition) fn(x) {
    return (x < 4);
} else fn(x) {
    return (x == 54);
};

pub const sort = fn(comptime T: type, arr: []T, f: Fn(T, T) bool) {
    // ...
};

pub const main = fn() void {
    var letters = []u8 {'g', 'e', 'r', 'm', 'a', 'n', 'i', 'u', 'm'};

    sort(u8, letters, fn(a, b) {
        return a < b;
    });
};
```

**Comment by locriacyber** [comment]
We should make the syntax easy to use for refactoring code. Inspiration: https://github.com/BSVino/JaiPrimer/blob/master/JaiPrimer.md#code-refactoring

Step 1
```
/// inline refactoring

const std = @import("std");

const V = struct {
    i: i32,
    u: i32,
};

pub fn main() !void {
    const v = V{.i=42, .u=35};
    std.log.debug("{}", .{v.i});
}
```

Step 2
```
pub fn main() !void {
    const v = V{.i=42, .u=35};
    {
        std.log.debug("{}", .{v.i});
    }
}
```

Step 3 (proposed syntax for block with restricted access to outer scope)
```
pub fn main() !void {
    const v = V{.i=42, .u=35};
    capture (v) {
        std.log.debug("{}", .{v.i});
    }
}
```

Step 4 (call inline anonymous function)
```
pub fn main() !void {
    const v = V{.i=42, .u=35};
    fn (v: V) {
        std.log.debug("{}", .{v.i});
    } (v);
}
```

Step 5 (extract function)
```
const foo = fn (v: V) {
    std.log.debug("{}", .{v.i});
};

pub fn main() !void {
    const v = V{.i=42, .u=35};
    foo(v);
}
```

**Comment by locriacyber** [comment]
I dislike the weird syntax of `.|x|`. A better solution is to make function definition
```zig
const foo = fn () void {};
```

This aligns with `const foo = struct {};`, and function definition is "function literal". Just like if `Struct` is a type, then `Struct{}` is constructor of that type, `fn () void` is a type, and `fn () void {}` is constructor of that type. With this syntax, nested function is as simple as
```zig
const foo = fn () void {
  const bar = fn () void {};
};
```

To be honest, I want to deprecate `fn functionName() {}`, because it makes anonymous function less discoverable. The language user then have to read the manual carefully to discover that function type is a feature.

**Comment by rohlem** [comment]
@locriacyber see #1717

**Comment by locriacyber** [comment]
I've tried to look into `zig/parser.zig` to add this syntax myself, but to no avail. Is there any other effort trying to implement this?

Also, I'm not sure what the syntax should be.

Closest to current syntax:
```
pub const foo = fn (a: i32) void {};
```

Separate type and implementation (like struct literal):
```
pub const foo = fn (i32) void |a| {};

pub const foo1: fn (i32) void = .|a| {};

const A: type =  fn (i32) void;
pub const foo2 = A |a| {};
```

**Comment by Vexu** [comment]
> Is there any other effort trying to implement this?

No effort has been made since this proposal has not been accepted and any work on it would likely end up being rejected.

**Comment by rohlem** [comment]
If syntax is a concern, here's an idea:
*Don't.
Just keep status quo.*

```zig
comptime {
  const foo1: fn (i32) void = fn _(a: i32) void {};

  const A: type =  fn (i32) void;
  pub const foo2 = @as(fn (i32) void, fn _(a: i32) void {});
  pub const foo3: A = fn _(a: i32) void {};

  //breaking news: now also at block scope
  fn fooriginal (a: i32) void {} // semicolon has to be a syntax error now that we allow hard tabs in files
}
```

Type inference can be a ~~rejected~~ follow-up proposal.

**Comment by Pyrolistical** [comment]
I think this proposal is a bit broad but a lower powered version would greatly improve common refactoring issues.
```zig

pub fn main() void {
  const a = A.init();
  defer a.deinit();

  const e1 = a.foo() + 10;
  const e2 = a.foo() + 100;
}
```

### Status quo, extract a struct
```zig
const ExtraStruct = struct {
  a: A,

  fn extracted(self: ExtraStruct, param: usize) usize {
     return self.a.foo() + param;
  }
};

pub fn main() void {
  const a = A.init();
  defer a.deinit();

  const extra_struct = ExtraStruct{.a = a};

  const e1 = extra_struct.extracted(10);
  const e2 = extra_struct.extracted(100);
}
```

Extracting a struct is very annoying, especially there are more common deps.

This proposal can help but we don't need its full power. Instead of an anonymous function, all that is needed here is an inlined anonymous parameterized block.

### Proposed, inlined anonymous parameterized block
```zig
pub fn main() void {
  const a = A.init();
  defer a.deinit();

  const extracted = inline blk: |param| {
    break :blk a.foo() + param;
  };

  const e1 = extracted(10);
  const e2 = extracted(100);
}
```

The idea is since `extracted` is comptime inlined, it is allowed access to lexically scoped variables.

Note that since `extracted` is a block and not a function, `return` would return to the outer function, just like normal blocks.

However, since it is a lexically scoped block, I would make it a compile error if it is passed like a closure.

```zig
fn foreach(self: @This(), closure: anytype) !void {
  for (self.buckets) |bucket| {
    var current = bucket.first;
    while (current) |node| : (current = node.next) {
       closure(&node.key, &node.value);
    }
  }
}

fn countValues(self: @This(), value: Value) usize {
    var count: usize = 0;
    const closure = inline blk: |_, v| {
       if (v.* == value) count += 1;
    };
    self.foreach(closure);  // compile error: `closure` not allowed to escape
    return count;
}
```
This mean it wouldn't work for https://github.com/ziglang/zig/issues/6965, where this example is adapted from.

**Comment by Skehmatics** [comment]
@Pyrolistical's proposal alone would be handy, but with a minor tweak this might sensibly solve [the open question](https://github.com/ziglang/zig/issues/1085#issuecomment-395974228) of how Zig can wrangle the semantics of C's function-like macros: **allow passing these to inline functions only.**

An example that has personally burned me a few times in the past is a particular procedural builder pattern used in C (ex, Pipewire POD, Clay). The idea is that you make a depth-stack builder for your structure, but then abstract away the push and pop to a macro that takes the middle as an argument. The most valuable thing about this pattern is not the abstraction, but that it puts the hierarchy you're creating front-and-center, which can vastly improve readability when that information is key. Compare these two, roughly modeled after Clay (https://github.com/nicbarker/clay/issues/3):
```c
OpenContainer(containerConfig); // Push a new entry to a hierarchy stack
PlainChild(config);             // Put an element into the container at the top of the hierarchy stack
PlainChild(config);
EndContainer(containerConfig);  // 2nd level
CompositeChild(config);
EndContainer();                 // Finalize 2nd level and pop it from the stack
for (int i = 0; i < tailLen; i++) { 
  PlainChild(tail[i]);
}
EndContainer();                 // and again for the 1st
```
```c
#define CONTAINER(config, children)
    OpenContainer(config);
    children
    CloseContainer();
...
CONTAINER(containerConfig, {
  PlainChild(config);
  PlainChild(config);
  CONTAINER(containerConfig, {
    CompositeChild(config);
  });
  for (int i = 0; i < tailLen; i++) { 
    PlainChild(tail[i]); 
  }
});
```

This type of abstraction is unrepresentable in status-quo Zig as far as I can tell, but fits very well with the ethos of making the most correct and performant approach the easiest. I'd argue that the degree it hides control flow is no less than that of async functions (not a good omen, to be fair) or labeled breaks. 
The example above under this modified proposal (type signature and definition syntax notwithstanding) would be:
```zig
inline fn CONTAINER(config: ContainerConfig, children: inline fn (void) void) void {
    OpenContainer(config);
    children();
    CloseContainer();
}

CONTAINER(containerConfig, inline {
  PlainChild(config);
  PlainChild(config);
  CONTAINER(containerConfig, inline {
    CompositeChild(config);
  });
  for (tail) |tailConfig| { 
    PlainChild(tailConfig); 
  }
});
```

## Alternatives
These come with a pretty big caveat: none of these can be done by translate-c, as the semantics of _calling_ must change to break through a layer of abstraction, so manual wrapping of the C library is required.

### Struct-based approach
Despite being the most straight forward seeming way to do this, it is not a zero-cost abstraction. Items were arranged merely by the sequential nature of execution previously, but now values must be returned so items can be accounted and re-combined in order. Additionally, this complicates inclusion of non-trivial logic, as you are now working with manipulating data rather than just writing statements.
```zig
CONTAINER(containerConfig, blk: { 
  var children: []Element = .{
    PlainChild(config),
    PlainChild(config),
    CONTAINER(containerConfig, .{
      CompositeChild(config),
    }),
  };
  for (tail) |tailConfig| { 
    children = children ++ .{ PlainChild(tailConfig) };
  }
  break :blk children;
});
```
### Defer and blocks
This is probably the biggest argument against this justification, as it *is* very explicit about what is happening without being entirely unreadable. However, if the pre- and post-condition are non-trivial (like for Pipewire), this can become untenable without adding extra abstraction elsewhere.
```zig
{ 
  OpenContainer(containerConfig); defer CloseContainer();
  PlainChild(config);
  PlainChild(config);
  {
    OpenContainer(containerConfig); defer CloseContainer();
    CompositeChild(config);
  }
  for (tail) |tailConfig| { 
    PlainChild(tailConfig); 
  }
}
